### PR TITLE
julia-mode: Move tests to new file and improve

### LINKED ISSUE #13237: julia-mode.el: Please don't define tests in the main library
I accidentally noticed that julia-mode.el defines its test right in the main library. 

It's very, very bad style, needlessly loads ERT for Julia mode files, and means that `M-x ert` will run Julia Mode tests as soon as Julia Mode was loaded, whether the user wants that or not.

No library should define its test inline.   Please, please don't do that, and move them into a separate test file instead.  Or at the very least, give me a flag to disable the tests.

### PR DESCRIPTION
**Pull request by justbur**
Create file julia-mode-tests.el to hold all ert tests, so that tests are
not loaded with julia-mode.el

Instead of ignoring-warnings for everyone to get around some users not
having ert, use require with the NOERROR flag.

Have julia--run-tests print a helpful message for thsoe who do not have
ert installed.

Fixes #13237

### CODE REVIEW

**Comment by yuyichao** [comment]
How should the test file be installed and run after this?

c.c.  @Wilfred

**Comment by justbur** [comment]
You can run them after loading the test file. I didn't see any scripts to
automate testing of julia-mode in the repo. How is it handled?

The point of this is to not load the tests for everyone who loads
julia-mode.el.

**Comment by yuyichao** [comment]
No the tests are not run on any CI AFAIK. I usually copy it locally and run the test. I was wondering if there are automatic hooks to tell emacs where to find the test script but loading the file manually is also fine.

**Comment by justbur** [comment]
With ert you can batch run the tests from the command line too.

I just realized there's a (require 'julia-mode) from the tests file. I'll
fix that later.

**Comment by justbur** [comment]
I rebased onto master and added the require along with a usage note for the command that you would need to run from the command line. Loading the file manually and running the tests is fine too.

File: contrib/julia-mode-tests.el [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+
+;; We can't use cl-lib whilst supporting Emacs 23 users who don't use
+;; ELPA.
+(with-no-warnings
+  (require 'cl)) ;; incf, decf, plusp
+
+(require 'julia-mode)
+
+;; Emacs 23.X doesn't include ert, so we only test if ert is available.
+(when (require 'ert nil t)
```
Comment (Wilfred) [inline contrib/julia-mode-tests.el]: The tests don't work on Emacs 23.X. Is it worth wrapping the entire file in this `when`? Now it's a separate test file, I'd be tempted to just do `(require 'ert)`

**Comment by Wilfred** [comment]
Looks good to me!

@yuyichao running the tests should just be a matter of opening the file, `M-x eval-buffer` then `M-x julia--run-tests` (or `M-x ert`).

File: contrib/julia-mode-tests.el (hunk 1 above)
Comment (justbur) [inline contrib/julia-mode-tests.el]: That's a good point. I don't think it needs to be wrapped anymore. 

I'll fix it now.

Comment (justbur) [inline contrib/julia-mode-tests.el]: Removed the when

**Comment by yuyichao** [comment]
LGTM and test passed locally (mainly to double check the parenthesis ;-p )

**Pull request merged**